### ISSUE TITLE
Garbage Collection?

### ISSUE THREAD

**Issue by napulen** [DDMAL/Neon#767]
Without reading much into it, there is this: https://github.com/WebAssembly/gc

It might be relevant for the memory leak problem found in Neon.

I'd start by looking at the Neon source code and seeing what operations in it could potentially be hoarding allocated memory.

**Comment by napulen** [comment]
Maybe this is not a memory leak problem but caching of different MEI file versions.

From @JRegimbal's presentation, it seems that whenever a `PUT` request fails, the intermediate MEI files are cached within the application (browser's cache?).

Recently, @JoyfulGen ran into an issue where Neon would take up to 4GB of RAM and be killed by the OS.

Maybe this memory consumption is related to the fallback routine? I think this only happens in the Web version of Neon.

**Comment by napulen** [comment]
The way I understand it, the `PUT` request will always fail in the gh-pages version of the app, right? Maybe I am misunderstanding the whole thing.

**Comment by JRegimbal** [comment]
@napulen no, you're correct. In the gh-pages version, it will always fail when a page is saved. It may be that this isn't saving to the database, but the cache entries or undo/redo stacks. The cache entry keeps a copy of every MEI file that's been loaded in memory, but shouldn't grow over time. The stacks maintain all the changes done in a session so undoing/redoing is possible and *will* grow.

**Comment by annamorphism** [comment]
I'm probably not understanding the issue correctly. Wouldn't the stacks have to contain on the order of thousands of MEI files/changes before they started hitting anything close to 4GB? I haven't tested, but I'd guess the actual problem is somewhere closer to 100 edits.  Maybe the pictures are also getting saved somewhere?

It seems to me a quick and dirty "solution" could simply be to only maintain the most recent ~10 changes in the session, and allow a "revert" to a saved state (in addition to reverting to the original upload.)

**Comment by napulen** [comment]
It'd be worth trying that "solution", if nothing else, as a way of debugging.

If the 10-change cache results in no memory growth/crashes in the browser, then we know that saving an unbounded number of intermediate files is related to the problem.

**Comment by jinh0** [comment]
@cadagong and I also agree that it must be an issue with cache. However, there are also some weird things that seem to bug out Neon, like drag selecting many syllables at once.

Just quickly testing the hypothesis with cache, I was able to get Neon to take up 98% of my RAM, 15.92 GB out of 16 GB of RAM by just spamming inserting a new element.

![Screen Shot 2022-05-30 at 12 40 03 PM](https://user-images.githubusercontent.com/40512164/171033991-abfafac3-55df-4fd2-89de-593725c701c6.png)

This means that it's not _just_ the issue with having too many MEI copies in cache. It's also a matter of us having to reload the entire SVG element every time a new change is made. Us having to reload the ENTIRE SVG at every change is equivalent to us reloading an entire website every time we make a change.

We think we can reduce the lag + memory consumption by choosing to reload only the parts of the MEI that have changed per editor action.

> Wouldn't the stacks have to contain on the order of thousands of MEI files/changes before they started hitting anything close to 4GB? 

The MEI files themselves may be very small in size, but our storage method, stored as both JavaScript objects and in PouchDB, may be requiring more memory per file.

This seems to be a multifaceted issue that requires further investigation. I will start looking into the issue.

**Comment by jinh0** [comment]
However, this is by no means the end of our problems regarding performance. Upon further examination, Neon suffers from a plethora of performance issues.

This is the rough list of CPU / RAM / performance issues I have collected:
## CPU Issues:
(@annamorphism confirmed that high CPU usage has been an ongoing issue!)
- [ ] Worker.js taking up 100% CPU usage
- [ ] Worker.js should be interrupted on new change
## Memory Issues:
- [ ] Firefox SVG duplication / garbage collection not working
- [ ] MEI strings take too much memory
## Performance Issues:
- [ ] Drag select with highlight lag
- [ ] updateHighlight lag
- [ ] Re-rendering entire SVG
- [ ] Parsing SVG string from Verovio

These are some critical issues regarding performance, but they require a difficult solution.

Long term issues are MEI strings taking too much memory and parsing SVG string / re-rendering entire SVG. Even with my memory leak fix, the sheer inefficiency of storing MEI strings and parsing increases RAM usage over time. Selective incremental re-rendering of the SVG is the clear solution, and for MEI strings, we should instead have an MEI abstract syntax tree that allows us to store small actions on the undo/redo stacks instead of the entire MEI string every time.

**Comment by napulen** [comment]
@JoyfulGen for completeness, could you indicate the latest memory consumption of the app after a 35-minute work session (i.e., after the fix)?

**Comment by napulen** [comment]
And congrats to all Neoners for squashing this one down.
